[WIP] Org hosts limit

jbradberry opened the pull request:
##### SUMMARY
This PR implements the feature allowing users to limit the number of hosts from a license an organization is allowed to use.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION
```
awx: 3.0.0
```


##### ADDITIONAL INFORMATION
- addition of a `max_hosts` field to the `Organization` model
- addition of an `org_host_limit_error` field to the `InventoryUpdate` model
- checks against an organization's limit now occur when
  - adding or changing hosts
  - starting jobs, job templates, workflow jobs, workflow job templates, and ad-hoc commands

jbradberry committed:
Added a max_hosts field to Organization

in order to optionally set a limit to the number of hosts an
organization that is sharing a license is allowed to manage.

related #1542

jbradberry committed:
Expose Organization.max_hosts in the API

jbradberry committed:
Include in the API the count of hosts used by an organization

jbradberry committed:
Update to fix tests

jbradberry committed:
Add permissions checks for the organization host limit

jbradberry committed:
Update the inventory_import management command

to respect the new Organization.max_hosts limit.

jbradberry committed:
Expose the new InventoryUpdate.org_host_limit_error field in the API

AlanCoding commented on awx/main/access.py:
```diff
@@ -838,6 +861,9 @@ def can_change(self, obj, data):
             self.check_license(add_host_name=data['name'])
 
+            # Check the per-org limit
+            self.check_org_host_limit(data, add_host_name=data['name'])
```
It's suspicious that this method is being called in `can_change` without `obj` being passed into it.

If I grok this correctly, we have to check when modifying the host, because the host might have previously been a duplicate name with another host, but then it changes to be a unique name, increasing the license impact. But it's also possible that it changes from a unique name to another unique name, and there's no way you can account for that without having `obj`. You might say that the same would apply to `self.check_license` above, and I would agree.

The org-specific check, however, has an additional problem that you won't be able to reliably determine the `inventory` field. It's possible that the user does a PATCH to change only the host `name`, and in that case IIRC, the `data` passed into `can_change` will only have the changing fields, meaning that you won't know the inventory, won't know the org, etc.

jbradberry replied:
It's definitely possible I got this wrong.  I'll take a closer look.

jbradberry replied:
I've changed it to a dictionary containing `org.inventory`, which if I'm following the reasoning correctly should be sufficient here.

AlanCoding commented on awx/api/views/mixin.py:
```diff
@@ -255,6 +257,8 @@ def get_serializer_context(self, *args, **kwargs):
             elif res == 'users':
                 org_reference = 'id'
+            elif res == 'hosts':
+                org_reference = 'inventory__organization'
```
oof, I know this logic is hard to understand, but looks like you extended it correctly. All of this count logic would be better if moved to the managers, but that's one of those things I thought of and never returned to. Implementing the new field via a manager method is great.

AlanCoding commented on awx/main/management/commands/inventory_import.py:
```diff
@@ -1032,7 +1064,6 @@ def handle(self, *args, **options):
             else:
                 tb = traceback.format_exc()
                 exc = e
-            transaction.rollback()
```
I don't follow this. In the event of a `CommandError`, the exception isn't re-raised, we do `sys.exit(1)` below, so I'd predict that it won't get rolled back if you remove this.

If an inventory update produced a license error, then it would be kind of inconvenient to not roll it back. That could put the user 1,000 hosts over license, and they'd have to go delete a bunch of content somewhere to get out of arrears.

jbradberry replied:
A `transaction.rollback()` is illegal when used with `transaction.atomic()`, which is already managing the rollback on exception behavior.  This can be seen if you add a `CommandError` above the line here (https://github.com/ansible/awx/blob/devel/awx/main/management/commands/inventory_import.py#L1005), and then run the `inventory_import` tests.

AlanCoding replied:
Don't we need to do something else to trigger the rollback then? Either use `set_rollback(True)` or raise (or re-raise) some database exception class?

jbradberry replied:
Because we are using `transaction.atomic()` as a context manager, any exception that is raised within that context will trigger the rollback.  The problem, if any, is whether we have bits of code here that need this protection but that are outside of the context manager.

AlanCoding replied:
> any exception that is raised within that context will trigger the rollback

but we caught the exception (this is something that the documentation says not to do). I don't actually see within this file where this code is wrapped in an atomic block. It's possible that the Django command class is already doing this, that's kind of implied by this code, but it's not a sure-fire bet that the existing code is correct either...

jbradberry replied:
I was a little bit fuzzy over whether it was only database exceptions or all exceptions which trigger the rollback, and I feel that Django's docs are a bit ambiguous in this regard, so I went looking:

https://github.com/django/django/blob/7e6b214ed34f5562dbd83cf54924a5b589a29715/tests/transactions/tests.py#L69-L74

It is all exceptions.

jbradberry replied:
The relevant protected code is here:

https://github.com/ansible/awx/blob/dd91920c5f858e3335054da3e74ae857724639b0/awx/main/management/commands/inventory_import.py#L1009-L1034

AlanCoding replied:
ok cool, so if I read correctly, the bottom line here seems to be that it will have already been rolled back by this point, and that line is old junk that didn't do anything?

jbradberry replied:
Yep, though worse than that, if it ever did get triggered it would have thrown another error preventing the more graceful bailout

jbradberry committed:
Correct the org limit check for changing hosts to use the host's org

instead of an inventory passed in from the user data, which is not allowed.

ryanpetrello commented on awx/main/migrations/0059_v350_organization_max_hosts.py:
```diff
@@ -0,0 +1,20 @@
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('main', '0058_v350_remove_limit_limit'),
```
FYI, decent chance this is going to merge soon: https://github.com/ansible/awx/pull/3186/files#diff-8a5cc29c38894e8e792e1b4172465788R9 and require you to update these migrations.

jbradberry replied:
yeah, figures.  I'll keep an eye on it.

ryanpetrello commented on awx/main/migrations/0060_v350_inventory_update_org_host_limit_error.py:
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
```
Might be nice to squash these two migration files together since it's one PR.

jbradberry replied:
Yeah, I'll do it when that other PR lands.

jbradberry committed:
Update the Organization API list and detail tests to check the host counts

matburt commented on awx/main/access.py:
```diff
@@ -320,6 +320,25 @@ def check_license(self, add_host_name=None, feature=None, check_expiration=True)
+
+        active_count = Host.objects.org_active_count(org.id)
+        if active_count > org.max_hosts:
+            raise PermissionDenied(_("Organization host limit of %s has been exceeded.") % org.max_hosts)
```
Should we leak the actual host limit in this way? I'm thinking we should not give the actual value here.

jbradberry replied:
Oh, hm, I hadn't really realized that reading an organization was restricted to the admin role.  Given that, I think I agree.

matburt commented on awx/main/models/organization.py:
```diff
@@ -39,6 +39,11 @@ class Meta:
         'InstanceGroup',
         blank=True,
     )
+    max_hosts = models.PositiveIntegerField(
```
This needs documentation (`help_text`) on the field definition so it gets picked up for our OPTIONS documentation.

See: https://github.com/ansible/awx/blob/devel/awx/main/models/inventory.py#L90

jbradberry replied:
Ok, I've added something now.

matburt reviewed (dismissed):
Just some small changes/discussion points.

jbradberry committed:
Remove the mention of the max_hosts value from the limit check messages

jbradberry committed:
Add test that AWX doesn't care about the limit when creating new hosts

jbradberry committed:
Add a test that AWX doesn't care about max_hosts when editing hosts

on an inventory.

jbradberry committed:
Tests for start permissions for JobTemplate and WorkflowJob

when max_hosts is set.

one-t commented:
Hey! I decided to fire up tower with this branch and hack on it a little- I've found that jobs cannot be launched in the current state.

`tower.log`
```
2019-02-12 23:42:59,853 WARNING  awx.api.generics status 400 received by user admin attempting to access /api/v2/job_templates/5/launch/ from 174.127.234.140
```

Response to POSTing to `$tower_url/api/v2/job_templates/$ID/launch`
```
Possibly unhandled rejection: {"data":{"detail":"Bad data found in related field inventory."},"status":400,"config":{"method":"POST","transformRequest":[null],"transformResponse":[null],"jsonpCallbackParam":"callback","url":"/api/v2/job_templates/5/launch/","headers":{"Accept":"application/json, text/plain, */*","X-Requested-With":"XMLHttpRequest","X-CSRFToken":"TMvvj6K1mB2ia7QO8RIhafrEFo2S6S2zd98whSbGRPbU3jJzVFiOLP9gTUONnFSX"}},"statusText":"Bad Request","xhrStatus":"complete"}
```

matburt reviewed (commented):
I have some comments mostly as a discussion for @wenottingham 

The following lines:
https://github.com/ansible/awx/pull/3174/files#diff-881951314805092fb39c41e94c4c92d3R336
https://github.com/ansible/awx/pull/3174/files#diff-45e09d8d8eee985479682c26e972f8e3R43

Imply that the primary use case is to limit the number of hosts for license reasons. If that's not the case then we should enforce `tower` source-types and per-org host limitations even under AWX.

If it is the case then we can carry on here.

wenottingham commented:
Yes, that description of the use case is correct.

jbradberry commented:
> Hey! I decided to fire up tower with this branch and hack on it a little- I've found that jobs cannot be launched in the current state.
> 
> `tower.log`
> 
> ```
> 2019-02-12 23:42:59,853 WARNING  awx.api.generics status 400 received by user admin attempting to access /api/v2/job_templates/5/launch/ from 174.127.234.140
> ```
> 
> Response to POSTing to `$tower_url/api/v2/job_templates/$ID/launch`
> 
> ```
> Possibly unhandled rejection: {"data":{"detail":"Bad data found in related field inventory."},"status":400,"config":{"method":"POST","transformRequest":[null],"transformResponse":[null],"jsonpCallbackParam":"callback","url":"/api/v2/job_templates/5/launch/","headers":{"Accept":"application/json, text/plain, */*","X-Requested-With":"XMLHttpRequest","X-CSRFToken":"TMvvj6K1mB2ia7QO8RIhafrEFo2S6S2zd98whSbGRPbU3jJzVFiOLP9gTUONnFSX"}},"statusText":"Bad Request","xhrStatus":"complete"}
> ```

Ugh, yeah ok, I think I see where I messed up.  I should be able to get a fix up pretty quickly.

jbradberry committed:
Properly use the inventory in the can_start permissions checks

jbradberry committed:
Add the new fields to the database

jbradberry committed:
Deal with the (erroneous) case where a job is missing the inventory

by bailing out of check_org_host_limit early.  Validation catches this
situation later on.

jbradberry committed:
Remove the hosts count from related_field_counts in the org api endpoints

It is probably not needed, and adds an additional db query.

AlanCoding commented:
I need to back up a little bit to the feature design, how will an organization admin find out in the UI what their current host count is vs. what their allocation is?

AlanCoding commented:
What roles are allowed to change `max_hosts` of an organization?

If an organization admin can just increase their max hosts, then it seems like that provides limited system protection, because there will be many users who can decide to consume more of the license allocation.

wenottingham commented:
> What roles are allowed to change `max_hosts` of an organization?

Sysadmin only.

jbradberry committed:
Restrict edit permissions on the Organization.max_hosts field to superusers

ryanpetrello commented on awx/api/serializers.py:
```diff
@@ -1265,6 +1265,20 @@ def get_summary_fields(self, obj):
+        obj_limit = getattr(obj, 'max_hosts', None)
+        api_limit = attrs.get('max_hosts')
+
+        if not view.request.user.is_superuser:
```
Yep, this makes sense 👍

jbradberry committed:
Update the permission error to include max_hosts and the current host count

jbradberry committed:
Change the wording of the error when adding a host

to "Organization host limit of %s would be exceeded...", since the
host will probably not actually be made active.

jbradberry committed:
Make the inventory_import command log an error message on exception

when it happens in the big try/except block in the middle of handle().
Previously it wasn't doing anything with it, except exiting with a
code of 1.

mabashian committed:
Add support for max hosts on org

unlikelyzero commented:
I am one of the two test owners assigned to this project. I'll be filing issues against this as I find them today.

mabashian committed:
Adds max value to host limit input

mabashian committed:
Default empty max hosts to 0

mabashian committed:
Adds error handling to launch api calls

mabashian committed:
Properly pass path to error message

AlexSCorey commented:
This addresses issue https://github.com/ansible/awx/issues/3290

mabashian committed:
Adds help text to host limit error shown in the output details of inventory sync jobs
